### PR TITLE
feat(ingestion) : Add Metabase Source Connector

### PR DESCRIPTION
**Pull request by jawadqu**
This PR adds the Metabase source connector.
Currently ingests Charts, dashboards, and associated metadata.
See [documentation](https://github.com/linkedin/datahub/blob/ea905000fa8390af5e20ec3f2abea7b944a9cab9/metadata-ingestion/source_docs/metabase.md) for more details.

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/source_docs/metabase.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Superset
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/metabase.md:1]: Metabase?

File: metadata-ingestion/source_docs/metabase.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+See documentation for Metabase's API at https://www.metabase.com/learn/administration/metabase-api.html 
+for more details on Metabase's login api.
+
+
+## Capabilities
+
+This plugin extracts the following:
+
+- Charts, dashboards, and associated metadata
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/metabase.md:17]: What about lineage from Charts to upstream Datasets?

File: metadata-ingestion/source_docs/metabase.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Check out the following recipe to get started with ingestion! See [below](#config-details) for full configuration options.
+
+For general pointers on writing and running a recipe, see our [main recipe guide](../README.md#recipes).
+
+```yml
+source:
+  type: metabase
+  config:
+    # Connection
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/metabase.md:29]: We tend to call this "Coordinates". Please keep this as such to maintain consistency

File: metadata-ingestion/source_docs/metabase.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+sink:
+  # sink configs
+```
+
+## Config details
+
+Note that a `.` is used to denote nested fields in the YAML recipe.
+
+| Field         | Required | Default            | Description                                             |
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/metabase.md:44]: Please mark the required configs.

File: metadata-ingestion/source_docs/metabase.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+| ------------- | -------- | ------------------ | ------------------------------------------------------- |
+| `connect_uri` |          | `"localhost:8088"` | Metabase host URL.                                      |
+| `username`    |          |                    | Metabase username.                                      |
+| `password`    |          |                    | Metabase password.                                      |
+| `env`         |          | `"PROD"`           | Environment to use in namespace when constructing URNs. |
+
+## Compatibility
+
+Coming soon!
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/metabase.md:53]: What version of Metabase did you validate against? Please put that here. Any new connector docs need this

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        )
+        title = dashboard_details.get("name", "") or ""
+        description = dashboard_details.get("description", "") or ""
+        last_modified = ChangeAuditStamps(
+            created=AuditStamp(time=modified_ts, actor=modified_actor),
+            lastModified=AuditStamp(time=modified_ts, actor=modified_actor),
+        )
+
+        card_urns = []
+        cards_data = dashboard_details.get("ordered_cards", "{}")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:141]: Nit: let's name this chart_urns

Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py:141]: We should aim to keep Metabase concepts named as they are in Metabase, but once they are mapped to DataHub concepts they should take on DataHub names

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+        return schema, name
+
+    @lru_cache(maxsize=None)
+    def get_datasource_from_id(self, datasource_id):
+        dataset_response = self.session.get(
+            f"{self.config.connect_uri}/api/database/{datasource_id}"
+        ).json()
+
+        return dataset_response["engine"], dataset_response["name"]
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:320]: Are we sure dataset_response["engine"] will correctly map to DataHub DataPlatform names? They are listed here: https://github.com/linkedin/datahub/blob/master/metadata-service/war/src/main/resources/boot/data_platforms.json

Are we sure that the dataset_response["name"] will map to the DataHub concept of name? Name should be fully qualified coordinates within a given platform. The exact structure can vary by platform.

File: metadata-ingestion/tests/integration/metabase/metabase_mces_golden.json [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                                "time": null,
+                                "actor": "urn:li:corpuser:admin@metabase.com",
+                                "impersonator": null
+                            },
+                            "deleted": null
+                        },
+                        "chartUrl": "http://localhost:3000/api/card/1",
+                        "inputs": [
+                            {
+                                "string": "urn:li:dataset:(urn:li:dataPlatform:h2,Sample Dataset.PUBLIC.PEOPLE,PROD)"
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/integration/metabase/metabase_mces_golden.json:77]: What is H2?

This Dataset name also looks off. 

Sample Dataset.PUBLIC.PEOPLE?

What does each part here represent?

File: metadata-ingestion/source_docs/metabase.md (hunk 2 above)
Comment (jjoyce0510) [reply metadata-ingestion/source_docs/metabase.md:17]: Let's note it explicitly.

File: metadata-ingestion/source_docs/metabase.md (hunk 1 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/source_docs/metabase.md (hunk 2 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 6 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py:145]: fixed please see [commit 0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab](https://github.com/linkedin/datahub/pull/3602/commits/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 7 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: fixed please see [commit 0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab](https://github.com/linkedin/datahub/pull/3602/commits/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab)

File: metadata-ingestion/tests/integration/metabase/metabase_mces_golden.json (hunk 8 above)
Comment (jawadqu) [reply metadata-ingestion/tests/integration/metabase/metabase_mces_golden.json]: fixed. Line 369+ in metabase.py
please see [commit 0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab](https://github.com/linkedin/datahub/pull/3602/commits/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab)

File: metadata-ingestion/source_docs/metabase.md (hunk 1 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/source_docs/metabase.md (hunk 2 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/source_docs/metabase.md (hunk 3 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/source_docs/metabase.md (hunk 4 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/source_docs/metabase.md (hunk 5 above)
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. Please see [Documentation](https://github.com/linkedin/datahub/blob/0c03ae4bc9ca7e473bb7c5c7f0ab6f3a1fb446ab/metadata-ingestion/source_docs/metabase.md)

**Review by jawadqu — COMMENTED** [review]
Hi John, all changes we dicussed have been checked in. Please review.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+
+    def construct_dashboard_from_api_data(
+        self, dashboard_info: dict
+    ) -> DashboardSnapshot:
+        dashboard_url = (
+            f"{self.config.connect_uri}/api/dashboard/{dashboard_info['id']}"
+        )
+        dashboard_response = self.session.get(dashboard_url)
+        dashboard_details = dashboard_response.json()
+        dashboard_urn = f"urn:li:dashboard:({self.platform},{dashboard_details['id']})"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:127]: would be better to use `builder.make_dashboard_urn`

ditto throughout - there's also make_chart_urn, make_user_urn, make_dataset_urn, etc.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+
+        engine_mapping = {
+            "sparksql": "spark",
+            "mongo": "mongodb",
+            "presto-jdbc": "presto",
+            "sqlserver": "mssql",
+            "bigquery-cloud-sdk": "bigquery",
+        }
+        if engine in engine_mapping:
+            platform = engine_mapping[dataset_response["engine"]]
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:367]:
```suggestion
            platform = engine_mapping[engine]
```

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+            "snowflake": "db",
+            "presto-jdbc": "catalog",
+            "presto": "catalog",
+            "mysql": "dbname",
+            "sqlserver": "db",
+        }
+
+        dbname = (
+            dataset_response.get("details", {}).get(
+                field_for_dbname_mapping[dataset_response["engine"]]
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:383]:
```suggestion
                field_for_dbname_mapping[engine]
```

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+            )
+            if engine in field_for_dbname_mapping
+            else None
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:386]: Are there any reasonable fallbacks here? Does it make sense to just try "db", "dbname", and "catalog" if the engine is unknown? 

If not, should we use `self.report.report_warning` to alert the end-user of an issue?

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py:386]: Similar question for the `platform` and `engine_mapping` above

**Review by hsheth2 — CHANGES_REQUESTED** [review]
Left a bunch of small comments but overall looks good!

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 9 above)
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py:127]: +1

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+        dataset_response = self.session.get(
+            f"{self.config.connect_uri}/api/database/{datasource_id}"
+        ).json()
+
+        # Map engine names to what datahub expects in
+        # https://github.com/linkedin/datahub/blob/master/metadata-service/war/src/main/resources/boot/data_platforms.json
+        engine = dataset_response["engine"]
+        platform = engine
+
+        engine_mapping = {
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:359]: nice!

File: metadata-ingestion/source_docs/metabase.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+- Title and description
+- Last edited by
+- Owner
+- Link to the dashboard in Metabase
+- Associated charts
+
+### Chart
+
+[/api/card](hhttps://www.metabase.com/docs/latest/api-documentation.html#card) endpoint is used to
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. see [metabase.md](https://github.com/linkedin/datahub/blob/ea905000fa8390af5e20ec3f2abea7b944a9cab9/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/source_docs/metabase.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+| Field                 | Required | Default            | Description                                             |
+| --------------------- | -------- | ------------------ | ------------------------------------------------------- |
+| `connect_uri`         |    ✅     | `"localhost:8088"` | Metabase host URL.                                      |
+| `username`            |    ✅     |                    | Metabase username.                                      |
+| `password`            |    ✅     |                    | Metabase password.                                      |
+| `database_alias_map`* |          |                    | Database name map to use when constructing dataset URN. |
+| `env`                 |          | `"PROD"`           | Environment to use in namespace when constructing URNs. |
+
+DataHub will try to determine database name from Metabase [api/database](https://www.metabase.com/docs/latest/api-documentation.html#database)
+payload. However, the name can be overridden from this map for a given database connected to Metabase.
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/source_docs/metabase.md]: fixed. see [metabase.md](https://github.com/linkedin/datahub/blob/ea905000fa8390af5e20ec3f2abea7b944a9cab9/metadata-ingestion/source_docs/metabase.md)

File: metadata-ingestion/tests/integration/metabase/test_metabase.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+                "com.linkedin.pegasus2avro.metadata.snapshot.ChartSnapshot"
+            ]["aspects"][0]["com.linkedin.pegasus2avro.chart.ChartInfo"][
+                "lastModified"
+            ][
+                "lastModified"
+            ][
+                "time"
+            ] = None
+
+            mce_helpers.assert_mces_equal(output, golden)
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/tests/integration/metabase/test_metabase.py]: fixed. Thanks for the awesome suggestion. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 17]
```diff
@@ -0,0 +1,407 @@
+from functools import lru_cache
+from typing import Dict, Iterable, Optional
+
+import dateutil.parser as dp
+import requests
+from pydantic.class_validators import validator
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: fixed. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+            {
+                "X-Metabase-Session": f"{self.access_token}",
+                "Content-Type": "application/json",
+                "Accept": "*/*",
+            }
+        )
+
+        # Test the connection
+        test_response = self.session.get(f"{self.config.connect_uri}/api/user/current")
+        if test_response.status_code == 200:
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: Good idea. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 9 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: done. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+            urn=dashboard_urn,
+            aspects=[],
+        )
+        last_edit_by = dashboard_details.get("last-edit-info") or {}
+        modified_actor = f"urn:li:corpuser:{last_edit_by.get('email', 'unknown')}"
+        modified_ts = int(
+            dp.parse(f"{last_edit_by.get('timestamp', 'now')}").timestamp() * 1000
+        )
+        title = dashboard_details.get("name", "") or ""
+        description = dashboard_details.get("description", "") or ""
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py:144]: Description can't be null. DataHub complained when I tried dashboard_details.get("description")

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 10 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: fixed. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 11 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: fixed. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py (hunk 12 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py:391]: We now report warning when we are unable to resolve platform or dbname. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)

**Review by jawadqu — COMMENTED** [review]
Implement Harshal's suggestions.

**Review by jjoyce0510 — APPROVED** [review]
Thanks @hsheth2 for all the great feedback and @jawadqu for effectively addressing everything. 

LGTM!

**Comment by ajaymuppuri** [comment]
When can we expect this source connector?

**Review by shirshanka — APPROVED** [review]
LGTM!

**Comment by jawadqu** [comment]
> 

@ajaymuppuri This has been merged. Please take a look and let us know if you would like anything improved.

**Pull request merged**